### ISSUE TITLE
Forward referenced functions not working.

### ISSUE THREAD

**Issue by gramster** [exaloop/codon#87]
I have some code which has two mutually recursive functions:

```python
def A():
   // under some condition
   B()

def B():
   // under some condition
   A()
```

This is perfectly legitimate Python, but Codon can't compile it, as it wants B to be declared already at the point it is called in A.

**Comment by inumanag** [comment]
Background: while Python binds variable names during the runtime, Codon does that during the compile time. Currently, our name analysis pass assumes that names are defined before being used. 

This limitation, however, will be lifted soon (probably the next release). There, we will analyze the whole module and  perform a look ahead for names.

**Comment by ghuls** [comment]
@inumanag 
Would this also allow static function variable like shown in:
https://www.adamsmith.haus/python/answers/how-to-declare-a-static-variable-in-a-function-in-python

**Comment by inumanag** [comment]
@ghuls We do not plan on supporting that for the time being. More generally, you cannot dynamically add slots to the existing objects in Codon---this is one of the current limitations that we are aware of but for which we have no ETA.

**Comment by ghuls** [comment]
> We do not plan on supporting that for the time being. More generally, you cannot dynamically add slots to the existing objects in Codon

Would it be possible to add a static variable to a function (static variable is defined at the same time as the function), similar to the __slots__ functionality for classes?

**Comment by lewisfogden** [comment]
I've found that you _can_ forward reference methods within a class, which may be a useable workaround:

```python
# Only Works in Python - can't forward reference functions in Codon (or declare an empty function / header)

def num_deaths(t: int) -> float:
    if t == 0:
        return 0
    else:
        return num_alive(t) * 0.01

def num_alive(t: int) -> float:
    if t == 0:
        return 1
    else:
        return num_alive(t-1) - num_deaths(t-1)

if __name__ == "__main__":
    print(f"{num_alive(10):.20}")

# Works in Both Codon and Python
class Model:
    def num_deaths(self, t: int) -> float:
        if t == 0:
            return 0
        else:
            return self.num_alive(t) * 0.01

    def num_alive(self, t: int) -> float:
        if t == 0:
            return 1
        else:
            return self.num_alive(t-1) - self.num_deaths(t-1)


if __name__ == "__main__":
    m = Model()
    print(f"{m.num_alive(10):.20}")
```

**Comment by inumanag** [comment]
Fixed in #20
